### DISCUSSION TITLE
Ozone without replication

### DISCUSSION THREAD

**Discussion by gigem** [apache/ozone#5563] [General]
I am trying to evaluate Ozone for use with Hive.  Because of IT decisions, our cluster runs on VMs with a SAN using RAID.  Consquently, we don't believe we need any replication from Ozone.  Unfortunately, I can't seem to get any working configuration without needing extra nodes.  Here is an example error:

2023-11-08 16:19:34,639 [RatisPipelineUtilsThread - 0] ERROR org.apache.hadoop.hdds.scm.SCMCommonPlacementPolicy: Unable to find enough nodes that meet the space requirement of 1073741824 bytes for metadata and 5368709120 bytes for data in healthy node set. Required 3. Found 1.

I've tried with the plain, docker-compose example (with --scale=1 and --scale=3) and also with an "onprem" installation using 1 node.  I've tried setting the default, replication type to STAND_ALONE and also with creating bickets with replication type set to RATIS and replication factor set to 1.  Strangely, I'm not allowed to create buckets with replication type set to STAND_ALONE.

I can't help but feel I'm missing something fundamental.  Can someone please guide me on how to configure Ozone for use without replication.  If that's not possible, how to do it with minimal waste.

**Comment by adoroszlai** [comment]
@gigem Can you please share which version of Ozone you are using?

**Reply by gigem** [reply]
Sure.  I'm using the prebuilt, version 1.3.0 tarball.  It looks like the Docker image is 20220623-1.

**Reply by adoroszlai** [reply]
Since 1.3.0 we have changed replication config in the docker-compose examples to fix them for single-node scenario (#5071).  I guess you ran into the same problem.

Can you please try replacing `ozone.replication=1` with `ozone.server.default.replication=1` in your config?

**Reply by gigem** [reply]
I will try shortly.  Thanks.

**Reply by gigem** [reply]
Still no joy.  I tried last with my build from git master on a single node system.  Our Hive partner did his own testing albeit with 3 nodes and scm/om running on of the datanodes.  He said the datanode needed a slightly different ozone-site.xml from scm/om so that is what I've also done.

Here is my ozone-site.xml used to start scm and om:
```
<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<configuration>
    <property>
        <name>ozone.om.address</name>
        <value>kbtest01</value>
    </property>
    <property>
        <name>ozone.metadata.dirs</name>
        <value>/data/meta</value>
    </property>
    <property>
        <name>ozone.scm.client.address</name>
        <value>kbtest01</value>
    </property>
    <property>
        <name>ozone.scm.names</name>
        <value>kbtest01</value>
    </property>
    <property>
      <name>ozone.scm.datanode.id.dir</name>
      <value>/data/meta/node</value>
    </property>
    <property>
      <name>ozone.default.bucket.layout</name>
      <value>FILE_SYSTEM_OPTIMIZED</value>
    </property>
    <property>
      <name>ozone.server.default.replication.type</name>
      <value>RATIS</value>
    </property>
    <property>
      <name>ozone.server.default.replication</name>
      <value>ONE</value>
    </property>
</configuration>
```

Here is my ozone-site.xml used to start datanode:
```
<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<configuration>
    <property>
        <name>ozone.om.address</name>
        <value>kbtest01</value>
    </property>
    <property>
        <name>ozone.metadata.dirs</name>
        <value>/data/datanode/meta</value>
    </property>
    <property>
        <name>ozone.scm.client.address</name>
        <value>kbtest01</value>
    </property>
    <property>
        <name>ozone.scm.names</name>
        <value>kbtest01</value>
    </property>
    <property>
      <name>ozone.scm.datanode.id.dir</name>
      <value>/data/meta/node</value>
    </property>
    <property>
      <name>hdds.datanode.dir</name>
      <value>/data/datanode/hdds</value>
    </property>
    <property>
      <name>ozone.default.bucket.layout</name>
      <value>FILE_SYSTEM_OPTIMIZED</value>
    </property>
    <property>
      <name>ozone.server.default.replication.type</name>
      <value>RATIS</value>
    </property>
    <property>
      <name>ozone.server.default.replication</name>
      <value>ONE</value>
    </property>
</configuration>
```

Finally, here the bucket configuration:
```
[ {
  "metadata" : { },
  "volumeName" : "hivemr3",
  "name" : "warehouse",
  "storageType" : "DISK",
  "versioning" : false,
  "usedBytes" : 4068,
  "usedNamespace" : 2,
  "creationTime" : "2023-11-08T21:30:18.393Z",
  "modificationTime" : "2023-11-08T21:30:18.393Z",
  "sourcePathExist" : true,
  "quotaInBytes" : -1,
  "quotaInNamespace" : -1,
  "bucketLayout" : "FILE_SYSTEM_OPTIMIZED",
  "owner" : "dengel",
  "link" : false
} ]
```

With this configuration, I still get errors like the following:
```
2023-11-08 21:32:15,911 [IPC Server handler 11 on default port 9863] ERROR org.apache.hadoop.hdds.scm.pipeline.WritableRatisContainerProvider: Unable to allocate a block for the size: 268435456, repConfig: RATIS/THREE
```

**Reply by glapark** [reply]
I am also trying to run Ozone with a single data node and with replication factor 1, in a local cluster (without using Docker or Kubernetes). I built Ozone using the commit in the master branch a couple of weeks ago.

With the default value of ozone.server.default.replication.type and ozone.server.default.replication, Ozone runs okay with 3 or 6 data nodes.

However, even with the following configuration, Ozone still expects 3 data nodes.

```
<property>
    <name>ozone.server.default.replication.type</name>
    <value>STAND_ALONE</value>
</property>

<property>
    <name>ozone.server.default.replication</name>
    <value>1</value>
</property>
```

`2023-11-09 10:15:38,149 [RatisPipelineUtilsThread-0] ERROR org.apache.hadoop.hdds.scm.SCMCommonPlacementPolicy: Unable to find enough nodes that meet the space requirement of 1073741824 bytes for metadata and 5368709120 bytes for data in healthy node set. Required 3. Found 1.`

So, my question is: does Ozone require 3 data nodes regardless of the configuration of ozone.server.default.replication?

**Reply by adoroszlai** [reply]
@glapark even if the default replication is set to 1, Ozone tries to create 3-node pipelines in the background.  Please ignore this error from `RatisPipelineUtilsThread`, it does not affect the ability to store data with replication 1.

**Reply by adoroszlai** [reply]
@gigem It turns out that, when using Ozone filesystem (as Hive does), setting server-side replication is not enough, we need to also set `ozone.replication=1` for the client.  This needs to be added to the `ozone-site.xml` in the same dir where `hdfs-site.xml` etc. are present.

**Reply by gigem** [reply]
Thanks for the follow up.  We use Docker images for Hive on Kubernetes so we'll have to investigate how to get the oxone-site.xml into the image.  In the meantime, I acquired 2 more VMs to use as datanodes so I can test with normal replication set to 3.

**Reply by adoroszlai** [reply]
You can also add it to `core-site.xml` or `hdfs-site.xml` if that's easier.

**Reply by gigem** [reply]
`core-site.xml` should be possible.  We have provisions for passing it into the Docker image.

**Reply by gigem** [reply]
Good news, finally!  I confirmed that adding `ozone.replication=1` to `core-site.xml` works with a single datanode.  I also previously confirmed that using only server-side configuration worked with 3 datanodes.

I now have one last question (I hope) for the near term.  Ozone 1.3.0 is quite old.  In fact, @glapark already identified 1 fairly recent fix that is needed to work with Hive.  My question then is which version of Ozone do you all recommend?  Are there any points on the master branch that are known to be reasonably stable?

**Reply by DaveTeng0** [reply]
hm, there'll be a 1.4.0 release around December!

**Reply by adoroszlai** [reply]
For dev testing and POC purposes I think latest commit from `master` is OK, with some caveats (known issues).  See details in on-going [discussion on `dev@` mailing list](https://lists.apache.org/list?dev@ozone.apache.org:2023-11:%22release%201.4.0%22) about next major release.

**Reply by gigem** [reply]
I've got about 1 or 2 weeks of POC testing left so a December release should work.  I have a couple more, hopefully quick questions.
- How well does fuse-dfs (or similar) work with Ozone?  In a few cases with non-S3/HDFS aware applications, we use a fuse mount to check file existance and copy data out.
- Is there a convenient way to copy/mirror data from S3 to Ozone?  We have serveral TBs of data we will want to move.

**Reply by adoroszlai** [reply]
@gigem I would like to suggest asking each of these questions in a separate discussion for better visibility (both to people who answer and those who are looking for answers).

**Reply by gigem** [reply]
@adoroszlai, done.

**Comment by DaveTeng0** [comment]
thanks @adoroszlai for the answer, that solves the same issue I got.
